### PR TITLE
[GH-1019] workload policies for create/dispatch and loading

### PR DESCRIPTION
**Pull request by ehigham**
### Purpose

Split from [GH-913]: External Exomes
The idea is to formalise the workload "interface" by the following methods: 

```
-- given a Transaction and a WorkloadRequest, create a new Workload
create-workload! :: Transaction -> WorkloadRequest -> Workload

-- given a Transaction and a Workload, start the Workload and return an updated snapshot
start-workload! :: Transaction -> Workload -> Workload

-- given a Transaction and a WorkloadRequest, return a stared Workload
execute-workload! :: Transaction -> WorkloadRequest -> Workload

-- given a Transaction and a Workload, update the workload's cromwell state
update-workload! :: Transaction -> Workload -> Workload
```

These are to be implemented in each module then the whole namespace is simply required in `handlers`

Also added the loading multimethod `load-workload-impl` - I need this for the exomes module but i'm including in this change for completeness.

### Testing

Refactored the endpoint tests into one parameterised system test namespace. I'm not sure how long should be an acceptable timeout for each workflow.

[GH-913]: https://broadinstitute.atlassian.net/browse/GH-913

### CODE REVIEW

File: api/src/wfl/api/workloads.clj [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  NOTE: do NOT call directly in product code - this is only meant to be called
+  within this namespace."
+  (fn [_ body] (:pipeline body)))
+
+(defn load-workload-for-uuid
+  "Use transaction `tx` to load `workload` with `uuid`."
+  [tx uuid]
+  (when-first [workload (jdbc/query tx
+                          ["SELECT * FROM workload WHERE uuid = ?" uuid])]
+    (load-workload-impl tx workload)))
```
Comment (rexwangcc) [inline api/src/wfl/api/workloads.clj:34]: like this, which will help simplify the aou module a lot.

**Review by rexwangcc — COMMENTED** [review]
Did a quick pass, looks great to me in general, like the interface implementation!

File: api/src/wfl/api/handlers.clj [hunk 2]
```diff
@@ -1,24 +1,32 @@
...
   [body]
-  (log/warn "Endpoint sent a 400 failure response:")
+  (log/warn "Endpoint returned 400 failure response:")
   (log/warn body)
   (-> body response/bad-request (response/content-type "application/json")))
 
+(defn fail-with-response
+  "A failure RESPONSE with BODY."
+  [response body]
+  (log/warn "Endpoint returned 404 failure response:")
```
[the remark this answers is not in the document]
Comment (ehigham) [reply api/src/wfl/api/handlers.clj:26]: Ah damn it. Edited the wrong function. Will add a test.

File: api/src/wfl/api/handlers.clj [hunk 3]
```diff
@@ -1,24 +1,32 @@
...
-              [ring.util.response :as response]
-              [wfl.module.aou :as aou]
-              [wfl.module.copyfile :as cp]
-              [wfl.jdbc :as jdbc]
-              [wfl.module.wgs :as wgs]
-              [wfl.service.cromwell :as cromwell]
-              [wfl.service.postgres :as postgres]
-              [wfl.wfl :as wfl]
-              [wfl.service.gcs :as gcs]))
+(ns wfl.api.handlers
```
[the remark this answers is not in the document]
Comment (ehigham) [reply api/src/wfl/api/handlers.clj:1]: No Idea. @jack-r-warren?

File: api/src/wfl/api/handlers.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehigham) [reply api/src/wfl/api/handlers.clj:26]: Good point. This was a silly mistake. Thanks for spotting it.

File: api/src/wfl/api/handlers.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (jack-r-warren) [reply api/src/wfl/api/handlers.clj:1]: Uh I have zero idea. I certainly didn't mean to add it. I'm really surprised no linter or anything every caught it.

[the remark this answers is not in the document]
Comment (rexwangcc) [reply api/src/wfl/api/handlers.clj:1]: No worries at all, just wanted to make sure it's not a special syntax of clojure I don't know yet..

File: api/src/wfl/module/aou.clj [hunk 4]
```diff
@@ -303,39 +286,17 @@
...
+  (->>
+    (add-aou-workload! tx request)
+    (workloads/load-workload-for-id tx)))
+
+(defmethod workloads/start-workload!
+  pipeline
+  [tx {:keys [id] :as workload}]
+  (do
+    (start-aou-workload! tx workload)
+    (workloads/load-workload-for-id tx id)))
```
Comment (jack-r-warren) [inline api/src/wfl/module/aou.clj:301]: I love this

**Review by jack-r-warren — APPROVED** [review]

**Pull request merged**